### ISSUE TITLE
Enum types are not checked in binary operators

### ISSUE THREAD

**Issue by JsonFreeman** [microsoft/TypeScript#3845]
I ran into this for real while trying to fix a bug in the compiler. We disallow assigning one enum to another if you use the `=` operator. But we do not disallow it for any of the compound assignment operators. Nor do we disallow for bitwise operators. I think we should disallow all of them.

``` ts
enum E { }
enum F { }

var e: E;
var f: F;

e = f; // Error
e |= f; // No error
var g = e | f; // No error, g is number
```

**Comment by mhegazy** [comment]
Looks like a regression in 1.1 from 1.0. this would be a breaking change now though.

**Comment by kemsky** [comment]
Currently it is possible to specify ordinal values:
```
enum AlertButtons {
    OK = 1,
    YES = 2,
    NO = 4,
    CANCEL = 8
}
```
Enums are already limited in many aspects (just compare with Java enums) , disallowing bitwise operators would make them even more useless.

**Comment by RyanCavanaugh** [comment]
@kemsky the proposal isn't to disallow bitwise operators on enums. It's to disallow bitwise operators between *different* enums.

**Comment by kemsky** [comment]
@RyanCavanaugh, i was not clear enough, my concern is only last line of example (it might have little sense, but in general there is nothing wrong):
```
var g = e | f; // No error, g is number
```

It is more important not to break the following statements:
```
var buttons = 1;
var isOk = (buttons | AlertButtons.OK) === AlertButtons.OK;
```

**Comment by RyanCavanaugh** [comment]
Operations between `number` and `enum` would still be allowed. The only thing to disallow is operators where both sides are `enum`s and they are not the same `enum` type.

**Comment by RyanCavanaugh** [comment]
I don't think this is coherent unless we were to patch the `enum <- number` assignability hole along with the `A | A` typing as `number` -- otherwise you get into an absurd situation where `(A | A) | B` is legal but `A | B` is illegal.
